### PR TITLE
Remove empty values from last-applied-configuration

### LINKED ISSUE #51: `nil` vs. `[]` vs. not-set-att-all
https://github.com/kubernetes/kubernetes/issues/70281

So if I send something either as `nil` or as in empty array, k8s API treats it as not set at all and omits the attribute on the resulting object. This causes problems now when k8s-client saves the config in annotation and determines the json patch operations based on that.

Apply following with k8s-client:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
  namespace: default
spec:
  selector:
    matchLabels:
      app: myapp
  template:
    metadata:
      labels:
        app: myapp
    spec:
      tolerations:
      containers:
      - name: myapp
        image: nginx:latest
        ports:
        - containerPort: 80
```

and after that try to apply e.g.:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
  namespace: default
spec:
  selector:
    matchLabels:
      app: myapp
  template:
    metadata:
      labels:
        app: myapp
    spec:
      tolerations: []
      containers:
      - name: myapp
        image: nginx:latest
        ports:
        - containerPort: 80
```

The json patch operations send would be:
```
[{"op": "replace", "path":"/spec/tolerations", "value":[]}]
```

Kube API borks with:
```
Error from server: jsonpatch replace operation does not apply: doc is missing key: /spec/tolerations
```

That's because the initial sending of "empty" `tolerations` results in the deployment object actually not having tolerations set at all so we cannot replace them either on later patches.

I think there's two ways to "fix" this:
1. "compact" the resource hash before storing it in the las-applied-config annotation. That way whatever was used omits nulls and empty arrays as if they were not given at all
2. Make the json patch part deal with these. If the patch "calculation" sees the old value as nil/empty array it could transform the operation from `replace` to `add`.

I _think_ option 1 might be more future proof and robust.

@jakolehm @kke WDYT?

### PR DESCRIPTION
**Pull request by jakolehm**
Fixes #51

### CODE REVIEW

File: lib/k8s/util.rb [hunk 1]
```diff
@@ -25,6 +25,20 @@ def self.compact_map(args)
...
 
+    # Recursive compact for Hash/Array
+    #
+    # @param hash_or_array [Hash,Array]
+    # @return [Hash,Array]
+    def self.recursive_compact(hash_or_array)
+      p = proc do |*args|
+        v = args.last
+        v.delete_if(&p) if v.respond_to?(:delete_if)
+        v.nil? || v.respond_to?(:"empty?") && v.empty?
```
Comment (kke) [inline lib/k8s/util.rb]:
```suggestion
        v.nil? || (v.respond_to?(:empty?) && v.empty?)
```

no need for quotes, parenthesis added for clarity.

File: lib/k8s/util.rb [hunk 2]
```diff
@@ -25,6 +25,20 @@ def self.compact_map(args)
...
+      end
+
+      hash_or_array.delete_if(&p)
```
Comment (kke) [inline lib/k8s/util.rb]: I don't see this working recursively, it only seems to do one level.

File: spec/k8s/util_spec.rb [hunk 3]
```diff
@@ -21,6 +21,27 @@
...
   end
 
+  describe '#recursive_compact' do
+    it 'compacts hashes recursively' do
+      expect(described_class.recursive_compact({
+        a: "a",
+        b: nil,
+        c: {
+          ca: "ca",
+          cb: nil
```
Comment (kke) [inline spec/k8s/util_spec.rb]:
```suggestion
          cb: nil,
          cc: [],
          cd: {}
```

spec deeper nesting

**Comment by jnummelin** [comment]
My biggest worry with this is on `NetworkPolicy` resources.  I.e. to allow all ingress traffic you'd use this kind of policy:
```
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: web-allow-external
spec:
  podSelector:
    matchLabels:
      app: web
  ingress:
  - from: []
```

In this case `from: []` would hit this with `:empty?`  being true...

Testing with kubectl by applying the above mentioned policy, the `kubectl.kubernetes.io/last-applied-configuration` does have:
```
"spec":{"ingress":[{"from":[]}]
```

**Comment by jakolehm** [comment]
Tried to match k8s logic more closely... PTAL.

**Pull request merged**